OpenRA.Game/Server/Connection.cs

```diff
@@ -195,9 +193,13 @@ void SendReceiveLoop(object s)
 			}
 		}
 
-		public void SendData(byte[] data)
+		public bool TrySendData(byte[] data)
 		{
+			if (sendQueue.IsAddingCompleted)
```

Comment:
This is still susceptible to a race condition since we've got multiple threads involved.

- The connection thread running `SendReceiveLoop` calls add, and `sendQueue.IsAddingCompleted` is false.
- The server thread drops a client and calls `sendQueue.CompleteAdding`.
- The connection thread calls `sendQueue.Add` and throws because the collection is marked as complete for adding.

Even `TryAdd` throws when the collection is marked completed for adding. Therefore I think a try-catch is needed here to be sure.

```c#
if (sendQueue.IsAddingCompleted)
    return false;

try
{
    sendQueue.Add(data);
    return true;
}
catch (InvalidOperationException)
{
    // Occurs if the collection is marked completed for adding by another thread.
    return false;
}
```

Reply:
Thanks. I'll add it for good measure.